scvi/inference/jvae_trainer.py

```diff
@@ -0,0 +1,386 @@
+
+    @property
+    def posteriors_loop(self):
+        return ["train_%d" % i for i in range(self.n_dataset)]
+
+    def data_loaders_loop(self):
+        posteriors = [self._posteriors[name] for name in self.posteriors_loop]
+        # find the largest dataset to cycle over the others
+        largest = np.argmax(
+            [posterior.gene_dataset.X.shape[0] for posterior in posteriors]
+        )
+
+        data_loaders = [
+            posterior if i == largest else cycle(posterior)
+            for i, posterior in enumerate(posteriors)
+        ]
+
+        return zip(*data_loaders)
+
+    def train(
+        self,
+        n_epochs: int = 20,
+        lr_d: float = 1e-3,
+        lr_g: float = 1e-3,
+        eps: float = 0.01,
+    ):
+        self.compute_metrics_time = 0
+        self.n_epochs = n_epochs
+
+        self.compute_metrics()
+        begin = time.time()
+
+        with torch.set_grad_enabled(True):
+            self.model.train()
+            self.discriminator.train()
+
+            d_params = filter(
+                lambda p: p.requires_grad, self.discriminator.parameters()
+            )
+            d_optimizer = torch.optim.Adam(d_params, lr=lr_d, eps=eps)
+
+            g_params = filter(lambda p: p.requires_grad, self.model.parameters())
+            g_optimizer = torch.optim.Adam(g_params, lr=lr_g, eps=eps)
+
+            train_discriminator = self.n_dataset > 1 and self.kappa_target > 0
+
+            with trange(
+                n_epochs, desc="training", file=sys.stdout, disable=self.verbose
+            ) as progress_bar:
+                for self.epoch in progress_bar:
+                    self.on_epoch_begin()
+                    progress_bar.update(1)
+                    for tensors in self.data_loaders_loop():
+                        if train_discriminator:
+                            latent_tensors = []
+                            for (i, (data, *_)) in enumerate(tensors):
+                                z = self.model.sample_from_posterior_z(
+                                    data, mode=i, deterministic=False
+                                )
+                                latent_tensors.append(z)
+
+                            # Train discriminator
+                            d_loss = self.loss_discriminator(
+                                [t.detach() for t in latent_tensors], True
+                            )
+                            d_loss *= self.kappa
+                            d_optimizer.zero_grad()
+                            d_loss.backward()
+                            d_optimizer.step()
+
+                            # Train generative model to fool discriminator
+                            fool_loss = self.loss_discriminator(latent_tensors, False)
+                            fool_loss *= self.kappa
+                            g_optimizer.zero_grad()
+                            fool_loss.backward()
```

Comment:
here I understand the kappa better and why you cut the optimization of the composite loss